MHArmory.Search/OpenCL/Host.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Collections.ObjectModel;
+using System.Text;
+using Cloo;
+
+namespace MHArmory.Search.OpenCL
+{
+    class Host : IDisposable
+    {
+        private ComputeContext Context { get; }
+        private ComputeProgram Program { get; }
+        private const string KernelName = "search";
+
+        public Host()
+        {
+            byte[] source = TrimUTFBOM(Properties.Resources.search);
+            string sourceStr = Encoding.ASCII.GetString(source);
+
+            // TODO: We need a proper platform / device selection mechanism.
+            ComputePlatform platform = ComputePlatform.Platforms[2];
+            var properties = new ComputeContextPropertyList(platform);
+            Context = new ComputeContext(ComputeDeviceTypes.All, properties, null, IntPtr.Zero);
+            Program = new ComputeProgram(Context, sourceStr);
+
+            var optionsBuilder = new HostOptionsBuilder();
+            optionsBuilder.AddDefine("MAX_RESULTS", SearchLimits.ResultCount);
+            optionsBuilder.AddDefine("MAX_DESIRED_SKILLS", SearchLimits.MaxDesiredSkills);
+            optionsBuilder.AddDefine("MAX_SET_SKILLS", SearchLimits.MaxSetSkills);
+            optionsBuilder.AddDefine("MAX_JEWEL_SIZE", SearchLimits.MaxJewelSize);
+            optionsBuilder.AddDefine("SLOTS_PER_EQUIPMENT", SearchLimits.SlotsPerEquipment);
+            optionsBuilder.AddDefine("SKILLS_PER_EQUIPMENT", SearchLimits.SkillsPerEquipment);
+            optionsBuilder.AddDefine("SET_SKILLS_PER_EQUIPMENT", SearchLimits.SetSkillsPerEquipment);
+            optionsBuilder.AddDefine("EQUIPMENT_TYPES", SearchLimits.EquipmentTypes);
+            string options = optionsBuilder.ToString();
+
+            Program.Build(null, options, null, IntPtr.Zero);
+        }
+
+        // For some reason when you embed a string to an assembly's resources, it adds a UTF BOM,
+        // and the OpenCL compiler doesn't like that. So this is to remove it manually.
+        private byte[] TrimUTFBOM(byte[] data)
+        {
+            byte[] bom = Encoding.UTF8.GetPreamble();
+            if (data.Length < bom.Length)
+            {
+                return data;
+            }
+
+            for (int i = 0; i < bom.Length; i++)
+            {
+                if (bom[i] != data[i])
+                {
+                    return data;
+                }
+            }
+
+            byte[] copy = new byte[data.Length - bom.Length];
+            Buffer.BlockCopy(data, bom.Length, copy, 0, copy.Length);
+            return copy;
+        }
+
+        public SerializedSearchResults Run(SerializedSearchParameters searchParameters)
+        {
+            ushort[] resultCount = new ushort[1];
+            const int resultLen = SearchLimits.ResultCount * (sizeof(ushort) * SearchLimits.TotalEquipments + 1 + 3 * (SearchLimits.TotalEquipments + 1) * SearchLimits.SlotsPerEquipment);
```

Comment:
I'd like to avoid all immediate values, please replace them all by constants, thus having names giving precious meaning to them. I recommend you to add a `Constants.cs` (or else) file in the `OpenCL` folder where you can store all your constants.